Add withastro/compiler project

jasikpark opened the pull request:
cc @natemoo-re

I'm currently trying to set up fuzz testing to augment the existing unit tests in https://github.com/withastro/compiler. Once this is accepted, I'll try to integrate CIFuzz for quick feedback on PRs

Using these guides for setup:
- https://google.github.io/oss-fuzz/getting-started/new-project-guide/
- https://google.github.io/oss-fuzz//getting-started/new-project-guide/go-lang/

I suppose I should try running the Dockerfile locally to see if it works?

jonathanmetzman commented:
Who uses withastro/compiler. Why is it a critical project that we should accept?
btw if you aren't accepted you can always use [clusterfuzzlite](https://github.com/google/clusterfuzzlite) instead of cifuzz/oss-fuzz

jasikpark commented:
It is a component of https://astro.build/, a web development framework

highest rated satisfaction meta-framework in the jamstack 2022 survey https://jamstack.org/survey/2022/#frameworks-by-usage-and-satisfaction (not expecting that to have context, but hopefully growing in usage)

Used by companies like Nord VPN - https://nordsecurity.com/blog/why-we-chose-astro-for-our-websites

Dunno what metric makes it critical 🤷 but it's well-used and I'm hoping to make the compiler well-robust!

https://ossinsight.io/analyze/withastro/astro

https://npmcharts.com/compare/@sveltejs/kit,@remix-run/react,@11ty/eleventy,astro,@redwoodjs/core,@shopify/hydrogen,@builder.io/qwik,@11ty/eleventy-plugin-webc?log=false&interval=7&minimal=true

![image](https://user-images.githubusercontent.com/10626596/208024054-f111efaf-d54e-4854-9a72-13cf21774d0d.png)

jasikpark commented:
Sorry, currently trying to debug locally what I've set up wrong..

getting a FUZZING_LANGUAGE not defined from /usr/local/bin/compile when running the docker container

jasikpark commented:
Current status after running `python infra/helper.py shell astro-compiler` and `compile` gives me:

```
root@621c50654301:/src/compiler# compile
---------------------------------------------------------------
Compiling libFuzzer to /usr/lib/libFuzzingEngine.a...  done.
---------------------------------------------------------------
CC=clang
CXX=clang++
CFLAGS=-O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link
CXXFLAGS=-O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link -stdlib=libc++
RUSTFLAGS=--cfg fuzzing -Zsanitizer=address -Cdebuginfo=1 -Cforce-frame-pointers
---------------------------------------------------------------
+ go install github.com/AdamKorcz/go-118-fuzz-build@latest
go: downloading github.com/AdamKorcz/go-118-fuzz-build v0.0.0-20221219200756-befc80a21334
go: downloading golang.org/x/tools v0.2.0
go: downloading golang.org/x/sys v0.2.0
+ go get github.com/AdamKorcz/go-118-fuzz-build/testing
go: downloading github.com/AdaLogics/go-fuzz-headers v0.0.0-20221206110420-d395f97c4830
go: downloading github.com/cyphar/filepath-securejoin v0.2.3
go: added github.com/AdaLogics/go-fuzz-headers v0.0.0-20221206110420-d395f97c4830
go: added github.com/AdamKorcz/go-118-fuzz-build v0.0.0-20221219200756-befc80a21334
go: added github.com/cyphar/filepath-securejoin v0.2.3
go: upgraded golang.org/x/sys v0.0.0-20221010170243-090e33056c14 => v0.2.0
+ compile_native_go_fuzzer github.com/withastro/compiler/internal FuzzTransformScoping fuzz_transform_scoping
go: downloading golang.org/x/net v0.0.0-20221004154528-8021a29435af
# command-line-arguments
./main.4061791584.go:31:9: undefined: astro.FuzzTransformScoping
2022/12/20 23:54:00 failed to build packages:exit status 2
root@621c50654301:/src/compiler#
```

jasikpark commented:
Reading https://github.com/google/oss-fuzz/blob/master/infra/base-images/base-builder/compile_native_go_fuzzer to understand where I'm erroring, seems that it is finding the function, but compiling w/in https://github.com/google/oss-fuzz/blob/35a9f36803ab3ebf19b19ce1f8f168c5b7be485c/infra/base-images/base-builder/compile_native_go_fuzzer#L18 is failing

jasikpark commented:
<img width="605" alt="image" src="https://user-images.githubusercontent.com/10626596/208790843-4dbb07fe-876f-49be-812e-0efac30ed1fe.png">
yesssssss

jasikpark commented:
Okay! Fuzzers should be compiling now

jasikpark commented:
https://google.github.io/oss-fuzz/advanced-topics/reproducing/#reproducing-build-failures

```
❯ python infra/helper.py check_build astro-compiler fuzz_transform_scoping
INFO:root:Running: docker run --rm --privileged --shm-size=2g --platform linux/amd64 -i -e FUZZING_ENGINE=libfuzzer -e SANITIZER=address -e ARCHITECTURE=x86_64 -e FUZZING_LANGUAGE=go -e HELPER=True -v /Users/calebjasik-defined/Github/oss-fuzz/build/out/astro-compiler:/out -t gcr.io/oss-fuzz-base/base-runner test_one.py fuzz_transform_scoping.
INFO: performing bad build checks for /tmp/not-out/tmpbf04yeqg/fuzz_transform_scoping
BAD BUILD: /tmp/not-out/tmpbf04yeqg/fuzz_transform_scoping seems to have either startup crash or exit:
/tmp/not-out/tmpbf04yeqg/fuzz_transform_scoping -rss_limit_mb=2560 -timeout=25 -seed=1337 -runs=4 < /dev/null
bash: line 1:    96 Killed                  /tmp/not-out/tmpbf04yeqg/fuzz_transform_scoping -rss_limit_mb=2560 -timeout=25 -seed=1337 -runs=4 < /dev/null
/usr/local/bin/bad_build_check: line 67:    18 Killed                  $FUZZER -seed=1337 -runs=$MIN_NUMBER_OF_RUNS &> $FUZZER_OUTPUT

ERROR:root:Check build failed.
```

Not sure what is dying here, I'd assume if this was a true fuzzing crash it'd be more verbose?

jonathanmetzman commented:
Not sure. Yeah looks more like an OOM.
We accept this project. Lemme know when this is ready to merge.

jasikpark commented:
is there a good debug command for this? Or just running `python infra/helper.py shell astro-compiler` then `compile` then replicating the command ran in `run_fuzzer`?

jonathanmetzman commented:
> is there a good debug command for this? Or just running `python infra/helper.py shell astro-compiler` then `compile` then replicating the command ran in `run_fuzzer`?

That's probably the best thing we've got sorry

jasikpark commented:
definitely an OOM, still gonna work on debugging why it's occurring.

jasikpark commented:
Maybe I'm just running into something local to my machine? Shouldn't the CI detect anything wrong w/ my setup?

jasikpark commented:
```
#231    NEW    cov: 62 ft: 70 corp: 5/15b lim: 6 exec/s: 0 rss: 34Mb L: 5/5 MS: 4 ChangeBit-CrossOver-EraseBytes-InsertRepeatedBytes-
#239    NEW    cov: 64 ft: 72 corp: 6/20b lim: 6 exec/s: 0 rss: 34Mb L: 5/5 MS: 3 ChangeByte-EraseBytes-CrossOver-
#273    NEW    cov: 66 ft: 74 corp: 7/26b lim: 6 exec/s: 0 rss: 34Mb L: 6/6 MS: 4 CopyPart-EraseBytes-EraseBytes-InsertRepeatedBytes-
#430    REDUCE cov: 66 ft: 74 corp: 7/25b lim: 6 exec/s: 0 rss: 34Mb L: 5/5 MS: 2 ChangeBit-EraseBytes-
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xc8 pc=0x6e2b56]

goroutine 17 [running, locked to thread]:
main.catchPanics()
        ./main.983543527.go:49 +0x35c
panic({0x8d5000, 0x9b8bb0})
        runtime/panic.go:884 +0x212
github.com/withastro/compiler/internal.PrintToSource(0x10c0001531f8, 0x0)
        github.com/withastro/compiler/internal/print-to-source.go:9 +0x76
github.com/withastro/compiler/internal/transform.FuzzTransformScoping.func1(0x577c8c?, {0x0, 0x0})
        github.com/withastro/compiler/internal/transform/transform_test.go_fuzz.go:171 +0x1c5
reflect.Value.call({0x8d0980?, 0x8eea50?, 0x13?}, {0x835c1f, 0x4}, {0x10c000201980, 0x2, 0x2?})
        reflect/value.go:584 +0x1bce
reflect.Value.Call({0x8d0980?, 0x8eea50?, 0x8379eb?}, {0x10c000201980, 0x2, 0x2})
        reflect/value.go:368 +0x1fa
github.com/AdamKorcz/go-118-fuzz-build/testing.(*F).Fuzz(0x10c000074d98, {0x8d0980?, 0x8eea50})
        github.com/AdamKorcz/go-118-fuzz-build@v0.0.0-20221222150057-fd2827626ad8/testing/f.go:177 +0x9ab
github.com/withastro/compiler/internal/transform.FuzzTransformScoping(0x16e?)
        github.com/withastro/compiler/internal/transform/transform_test.go_fuzz.go:163 +0x97
main.LibFuzzerFuzzTransformScoping({0x602000006250, 0x8, 0x8})
        ./main.983543527.go:31 +0x133
main.LLVMFuzzerTestOneInput(0x602000006250, 0x8)
        ./main.983543527.go:24 +0xa5
==2296== ERROR: libFuzzer: deadly signal
    #0 0x539071 in __sanitizer_print_stack_trace /src/llvm-project/compiler-rt/lib/asan/asan_stack.cpp:87:3
    #1 0x457b88 in fuzzer::PrintStackTrace() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerUtil.cpp:210:5
    #2 0x43c863 in fuzzer::Fuzzer::CrashCallback() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:233:3
    #3 0x7fc4874a741f  (/lib/x86_64-linux-gnu/libpthread.so.0+0x1441f) (BuildId: 7b4536f41cdaa5888408e82d0836e33dcf436466)
    #4 0x5cf700 in runtime.raise.abi0 runtime/sys_linux_amd64.s:158

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 5 InsertRepeatedBytes-ChangeBit-ChangeBinInt-ChangeBinInt-CMP- DE: "\000\000\000\000\000\000\0006"-; base unit: 7f878b38b3bddd398b5a23144939907eb973fa81
0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x36,
\000\000\000\000\000\000\0006
artifact_prefix='./'; Test unit written to ./crash-f00e356e375e0ee567c6e2cad77f4262d989410d
Base64: AAAAAAAAADY=
root@eaa1e2f637a3:/src/compiler# 
```
lul i gave the docker image too little memory or something on mac - it runs fine in <https://gitpod.io>

@jonathanmetzman I believe the initial harness is ready for merge in this case

jasikpark commented:
still gotta fix some immediate bugs that the FuzzScopeHTML fuzzer has found 😓, but I imagine that's not required for merging!

jasikpark commented on projects/astro-compiler/build.sh:
```diff
@@ -0,0 +1,24 @@
+################################################################################
+
+# Fuzz internal/transform
+printf "package transform\nimport _ \"github.com/AdamKorcz/go-118-fuzz-build/testing\"\n" > $SRC/compiler/internal/transform/registerfuzzdep.go
```
used this cause i saw it in other go projects, not sure it's best practices but it works!

jasikpark commented on projects/astro-compiler/build.sh:
```diff
@@ -0,0 +1,24 @@
+go get -u github.com/AdamKorcz/go-118-fuzz-build/testing
+
+compile_native_go_fuzzer $SRC/compiler/internal/transform FuzzScopeHTML fuzz_scope_html
+compile_native_go_fuzzer $SRC/compiler/internal/transform FuzzTransformScoping fuzz_transform_scoping
```
I'm excited to make this list grow and get them all to long-running w/ low crashes!

jonathanmetzman reviewed (approved):
lgtm

jasikpark commented:
wahoo
